library/core/src/fmt/mod.rs

```diff
@@ -1693,49 +1693,37 @@ impl<'a> Formatter<'a> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn pad(&mut self, s: &str) -> Result {
-        // Make sure there's a fast path up front
+        // Make sure there's a fast path up front.
         if self.options.width.is_none() && self.options.precision.is_none() {
             return self.buf.write_str(s);
         }
-        // The `precision` field can be interpreted as a `max-width` for the
+
+        // The `precision` field can be interpreted as a maximum width for the
         // string being formatted.
-        let s = if let Some(max) = self.options.precision {
-            // If our string is longer that the precision, then we must have
-            // truncation. However other flags like `fill`, `width` and `align`
-            // must act as always.
-            if let Some((i, _)) = s.char_indices().nth(max) {
-                // LLVM here can't prove that `..i` won't panic `&s[..i]`, but
-                // we know that it can't panic. Use `get` + `unwrap_or` to avoid
-                // `unsafe` and otherwise don't emit any panic-related code
-                // here.
-                s.get(..i).unwrap_or(s)
-            } else {
-                &s
-            }
+        let max_char_count = self.options.precision.unwrap_or(usize::MAX);
+        let mut iter = s.chars();
+        let char_count = iter.by_ref().take(max_char_count).count();
```

Comment:
`s.chars().count()` has a special implementation (see https://doc.rust-lang.org/1.84.0/src/core/str/iter.rs.html#46 and https://doc.rust-lang.org/1.84.0/src/core/str/count.rs.html) that is much faster than iterating through all the chars one by one.

`s.chars().by_ref().take(_).count()` on the other hand goes through the `Take` adapter, which does not use this optimized counting implementation.

This means that with this change, even a simple `println!("{}", some_string)` will now pull in the code for UTF8 decoding, blowing up binary size. And it'll be a bit slower, too.

Reply:
Though perhaps `s.char_indices().nth(max)` already pulled in that code anyway. That's still an opportunity for improvement.

Regardless, we should use the optimized char counting algorithm when there is no precision (max length).

Reply:
Since `Chars` has an optimized implementation of `Iterator::advance_by`, but `CharIndices` does not, I added an implementation to take advantage of that. This has the nice benefit of not bumping the offset as an induction variable, which I was trying to avoid manually before by just using `Chars`.

Using `.nth()` introduces an unnecessary `.next()` and `.take(n).count()` does a fold and iterates by one, so I use `.char_indices().advance_by(n)`, which I think does the minimum possible with this API.

I've changed it to switch between `.char_indices().advance_by(n)` and `.chars().count()`, depending on whether truncation is needed. It would be interesting to benchmark `.char_indices().advance_by(usize::MAX)` against `.chars().count()`.